hamilton/function_modifiers/macros.py

```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
+
+        To indicate that a node should share a namespace with the rest of the apply(...) operations in a pipe, one can pass in `...` (the ellipsis).
+
+        .. code-block:: python
+
+            @pipe(
+                apply(_add_one).named("a", namespace="foo"), # foo.a
+                apply(_add_two, y=source("upstream_node")).named("b", namespace=...), # final_result.b
+            )
+            def final_result(upstream_int: int) -> int:
+                return upstream_int
+
+        Note that if you pass a namespace argument to the `pipe` function, it will set the namespace on each sub-apply operation.
+        This is useful if you want to ensure that all the nodes in a pipe have a common namespace, but you want to rename them.
+
+        .. code-block:: python
+
+            @pipe(
+                apply(_add_one).named("a"), # a
+                apply(_add_two, y=source("upstream_node")).named("b"), # foo.b
+                namespace=..., # default -- final_result.a and final_result.b, OR
+                namespace=None, # no namespace -- a and b are exposed as that, OR
+                namespace="foo", # foo.a and foo.b
+            )
+            def final_result(upstream_int: int) -> int:
+                return upstream_int
+
+        In all likelihood, you should not be using this, and this is only here in case you want to expose a node for
+        consumption/output later. Setting the namespace in individual nodes as well as in `pipe` is not yet supported.
+    """
+
+    def __init__(
+        self, *transforms: Applicable, namespace: NamespaceType = ..., collapse=False, _chain=False
+    ):
+        """Instantiates a `@pipe` decorator.
+
+        :param transforms: apply transformation to be applied, in order
+        :param namespace: namespace to apply to all nodes in the pipe. This can be "..." (the default), which resolves to the name of the decorated function, None (which means no namespace), or a string (which means that all nodes will be namespaced with that string). Note that you can either use this *or* namespaces inside pipe()...
+        :param collapse: Whether to collapse this into a single node. This is not currently supported.
+        :param _chain: Whether to chain the first parameter. This is the only mode that is supported. Furthermore, this is not externally exposed. @flow will make use of this.
+        """
+        self.transforms = transforms
+        self.collapse = collapse
+        self.chain = _chain
+        self.namespace = namespace
+
+        if self.collapse:
+            raise NotImplementedError(
+                "Collapsing apply() functions as one node is not yet implemented for pipe(). Please reach out if you want this feature."
+            )
+
+        if self.chain:
+            raise NotImplementedError("@flow() is not yet supported -- this is ")
+
+    def inject_nodes(
```

Comment:
this function is too long, with too much logic? will be easier to test if it was smaller functions.

Reply:
Meh it doesn't break nicely into smaller functions that need testing, and testing the smaller functions might be a little better for correctness, but would likely serve to lock down a contract that isn't what we want, and slow future development.

I'd rather have one larger function that has quite a few pieces that will change, with a clear contract and tests on it, then break it up purely for the sake of having higher test coverage.

Reply:
i dont' want to read it is the other feedback 😛